src/persist-client/src/cache.rs

```diff
@@ -250,10 +262,179 @@ async fn consensus_rtt_latency_task(
+/// command for the same shard are executing concurrently, only one can win
+/// anyway, the other will retry. With the mutex, we even get to avoid the retry
+/// if the racing commands are on the same process.
+#[derive(Debug, Default)]
+pub struct StateCache {
+    states: Mutex<BTreeMap<ShardId, Arc<OnceCell<Weak<dyn DynState>>>>>,
+}
+
+#[derive(Debug)]
+enum StateCacheInit {
+    Init(Arc<dyn DynState>),
+    NeedInit(Arc<OnceCell<Weak<dyn DynState>>>),
+}
+
+impl StateCache {
+    pub(crate) async fn get<K, V, T, D, F, InitFn>(
+        &self,
+        shard_id: ShardId,
+        mut init_fn: InitFn,
+    ) -> Result<Arc<Mutex<TypedState<K, V, T, D>>>, Box<CodecMismatch>>
+    where
+        K: Debug + Codec,
+        V: Debug + Codec,
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64,
+        F: Future<Output = Result<TypedState<K, V, T, D>, Box<CodecMismatch>>>,
+        InitFn: FnMut() -> F,
+    {
+        loop {
+            let init = {
+                let mut states = self.states.lock().await;
+                let state = states.entry(shard_id).or_default();
+                match state.get() {
+                    Some(once_val) => match once_val.upgrade() {
+                        Some(x) => StateCacheInit::Init(x),
+                        None => {
+                            // If the Weak has lost the ability to upgrade,
+                            // we've dropped the State and it's gone. Clear the
+                            // OnceCell and init a new one.
+                            *state = Arc::new(OnceCell::new());
+                            StateCacheInit::NeedInit(Arc::clone(state))
+                        }
+                    },
+                    None => StateCacheInit::NeedInit(Arc::clone(state)),
+                }
+            };
+
+            let state = match init {
+                StateCacheInit::Init(x) => x,
+                StateCacheInit::NeedInit(init_once) => {
+                    let mut did_init: Option<Arc<Mutex<TypedState<K, V, T, D>>>> = None;
+                    let state = init_once
+                        .get_or_try_init::<Box<CodecMismatch>, _, _>(|| async {
+                            let init_res = init_fn().await;
+                            let state = Arc::new(Mutex::new(init_res?));
+                            let ret = Arc::downgrade(&state);
+                            did_init = Some(state);
+                            let ret: Weak<dyn DynState> = ret;
+                            Ok(ret)
+                        })
+                        .await?;
+                    if let Some(x) = did_init {
+                        // We actually did the init work, don't bother casting back
+                        // the type erased and weak version.
+                        return Ok(x);
+                    }
+                    let Some(state) = state.upgrade() else {
+                        // WIP
```

Comment:
is there anything else to do here? offhand continuing seems like the only option

Reply:
ha, whoops! looked at this again and I'm pretty sure this was a "write a comment explaining what's going on here" sort of WIP, so filled that in